corehq/apps/reports/tests/test_views.py

```diff
@@ -0,0 +1,578 @@
+from django.core.exceptions import PermissionDenied
+from django.test import TestCase, RequestFactory
+from unittest.mock import patch
+from corehq.apps.domain.shortcuts import create_domain
+from corehq.apps.users.models import Permissions, UserRole, WebUser
+from corehq.apps.saved_reports.models import ReportConfig, ReportNotification
+from corehq.blobs import get_blob_db
+import unittest
+
+from .. import views
+
+REPORT_NAME_LOOKUP = {
+    'worker_activity': 'corehq.apps.reports.standard.monitoring.WorkerActivityReport'
+}
+
+
+class TestEmailReport(TestCase):
+    ARG_INDEX = 0
+    RECIPIENT_INDEX = 0
+
+    @patch.object(views, 'send_email_report')
+    def test_user_can_send_email(self, mock_send_email):
+        self._set_user_report_access(REPORT_NAME_LOOKUP['worker_activity'])
+        response = self.email_report(report_name='worker_activity')
+        mock_send_email.delay.assert_called_once()
+        self.assertEqual(response.status_code, 200)
+
+    @patch.object(views, 'send_email_report')
+    def test_email_is_sent_to_recipient_list(self, mock_send_email):
+        # NOTE: It appears the email form is broken and will only accept a single email
+        # so that's all that is being tested here
+        self.request = self._create_request(params={'recipient_emails': 'user1@test.com'})
+        self.email_report()
+
+        self.assertEqual(
+            mock_send_email.delay.call_args[self.ARG_INDEX][self.RECIPIENT_INDEX],
+            {'user1@test.com'}
+        )
+
+    @patch.object(views, 'send_email_report')
+    def test_owner_flag_adds_owner_email_to_recipient_list(self, mock_send_email):
+        # Setup creates a user with email address 'test_user@dimagi.com'
+        self.request = self._create_request(params={
+            'recipient_emails': 'user1@test.com',
+            'send_to_owner': True
+        })
+        self.email_report()
+
+        self.assertEqual(
+            mock_send_email.delay.call_args[self.ARG_INDEX][self.RECIPIENT_INDEX],
+            {'test_user@dimagi.com', 'user1@test.com'}
+        )
+
+    def test_invalid_form_returns_bad_request(self):
+        invalid_form_params = {}
+        self.request = self._create_request(params=invalid_form_params)
+        response = self.email_report()
+        self.assertEqual(response.status_code, 400)
+
+    def test_invalid_slug_returns_404(self):
+        with self.assertRaises(Http404):
+            self.email_report(report_name='nonexistant_report')
+
+    def test_when_user_cannot_view_report_receives_404(self):
+        self._set_user_report_access('not_this_report')
+        with self.assertRaises(Http404):
+            self.email_report(report_name='worker_activity')
+
+# ################ Helpers / Setup
+    def email_report(self, domain=None, report_name='worker_activity'):
+        domain = domain or self.domain
+        return views.email_report(self.request, domain, report_name, once=True)
+
+    @classmethod
+    def setUpClass(cls):
```

Comment:
I've come to prefer the `setUpTestData` class method because it doesn't require overriding a super method and it is executed inside the main test transaction so any SQL object created by it are cleaned up if there is an error.

Reply:
Once we've migrated off Couch, I like this suggestion. Until that point, however, I'm concerned this approach adds additional complexity. Because couch objects require a manual deletion, and because the `unittest` module provides no class-based `addClassCleanup()`, we're forced to handle manual deletion in `tearDownClass()`. If I'm forced to use `tearDownClass()`, I expect to see the corresponding creation in `setUpClass()`, not `setUpTestData()`.

Reply:
That just seems like a conventional expectation (which could be changed). Fine to leave as is but I still think `setUpTestData` is a better option.